src/crop-manager.sol

```diff
@@ -0,0 +1,171 @@
+    function hope(address) external;
+    function nope(address) external;
+}
+
+interface CropLike {
+    function gem() external view returns (address);
+    function ilk() external view returns (bytes32);
+    function join(address, address, uint256) external;
+    function exit(address, address, uint256) external;
+    function tack(address, address, uint256) external;
+    function flee(address) external;
+}
+
+interface TokenLike {
+    function approve(address, uint256) external;
+    function transferFrom(address, address, uint256) external;
+}
+
+contract UrnProxy {
+    address immutable public vat;
+    address immutable public usr;
+    address public owner;
+
+    constructor(address vat_, address usr_) public {
+        owner = msg.sender;
+        vat = vat_;
+        usr = usr_;
+        VatLike(vat_).hope(msg.sender);
+    }
+
+    function migrate(address newOwner) external {
+        require(msg.sender == owner, "UrnProxy/not-owner");
+        VatLike(vat).hope(newOwner);
+        VatLike(vat).nope(owner);
+        owner = newOwner;
+    }
+}
+
+contract CropJoinManager {
+    mapping (address => uint256) public wards;
+    mapping (address => address) public proxy;  // UrnProxy per user
+    address public implementation;
+
+    event Rely(address indexed);
+    event Deny(address indexed);
+    event SetImplementation(address indexed);
+
+    constructor() public {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    function rely(address usr) public auth { wards[usr] = 1; emit Rely(msg.sender); }
+    function deny(address usr) public auth { wards[usr] = 0; emit Deny(msg.sender); }
+    modifier auth { require(wards[msg.sender] == 1, "CropJoinManager/non-authed"); _; }
+
+
+    function setImplementation(address implementation_) external auth {
+        implementation = implementation_;
+        emit SetImplementation(implementation_);
+    }
+
+    fallback() external {
+        address _impl = implementation;
+        require(_impl != address(0));
+
+        assembly {
+            let ptr := mload(0x40)
+            calldatacopy(ptr, 0, calldatasize())
+            let result := delegatecall(gas(), _impl, ptr, calldatasize(), 0, 0)
+            let size := returndatasize()
+            returndatacopy(ptr, 0, size)
+
+            switch result
+            case 0 { revert(ptr, size) }
+            default { return(ptr, size) }
+        }
+    }
+}
+
+contract CropJoinManagerImp {
+    bytes32 slot0;
+    mapping (address => address) public proxy;  // UrnProxy per user
+
+    address public immutable vat;
+    constructor(address vat_) public {
+        vat = vat_;
+    }
+
+    function getProxy(address usr) internal returns (address urp) {
+        urp = proxy[usr];
+        if (urp == address(0)) {
+            urp = proxy[usr] = address(new UrnProxy(address(vat), usr));
+        }
+    }
+
+    function createProxy() external returns (address) {
+        return getProxy(msg.sender);
+    }
```

Comment:
I'd prefer a single `getOrCreateProxy` function. If you're concerned about saving gas internally, I'd just duplicate code, tbh. If the name is too long I can think of a number of cute memes, e.g. "create" + "get" = "cret" ==> `cretProxy` (REST lovers cringing).

Reply:
Let's discuss this on the call. I don't have strong opinions besides I think it is useful to allow just creating a proxy without the wasted gas of a `join(..., 0)`.

Reply:
Conclusion: merge or at least fix naming.

Reply:
Done.